extensions/net/01_net.js

```diff
@@ -207,7 +207,16 @@
       ...options,
     });
 
-    return new Listener(res.rid, res.localAddr);
+    const listener = new Listener(res.rid, res.localAddr);
+    if (options?.signal) {
+      // TODO(benjamingr) there is memory leak potential here if the signal outlives the
+      // server significantly it can retain it. Ideally Deno would have internal weak
+      // event listener like Node does *or* it would have an event for the server closing.
+      options?.signal?.addEventListener("abort", () => listener.close(), {
+        once: true,
+      });
+    }
```

Comment:
```suggestion
    if (options?.signal) {
      const weak = new WeakRef(listener);
      options?.signal.addEventListener("abort", () => weak.deref()?.close(), {
        once: true,
      });
    }
```

Reply:
That won't help - it still leaks the listener :) Happy to elaborate if that is unclear

Reply:
it's true that function captures everything
if v8 is failing to optimize that without any invocations, we can move event callback creation outside scope

Reply:
@evanwashere even if the capture was only of the listener this is still a memory leak since the listener itself is captured.

Weak event listeners need to be baked into the event target implementation itself for this to not leak (here is an example how I did it in Node.js https://github.com/nodejs/node/pull/36607 and I am happy to point to whatwg discussions).

That said - another (not bad) alternative is to expose a way to know when the listener is closed (and remove the handler then) but I didn't want to propose a change of the API of Deno.listen just for that :] 

(This leak is small and not very problematic to be fair)

Reply:
listener is wrapped in weakref, so capture won't matter and problem will be fixed

```js
{
  const listener = {...};
  if (signal) signal.add('event', createListenerAbortCallback(new WeakRef(listener)));
}

function createListenerAbortCallback(weakref) {
  return () => weakref.deref()?.close();
}
```

Reply:
but the signal would still have `() => weakref.deref()?.close();` in its listener array and that function would never gets cleared.

Reply:
that's a fundamental problem with event listeners and needs a separate fix
the goal right now is to prevent signal from retaining hard reference to listener object and other temporary objects

Reply:
yes, the fact it needs a separate fix is why I added a TODO for it :) The actual leak here is contrived, edge-casey and not that bad.

Reply:
Bit of a drive-by comment but can't this be fixed with a WeakMap?
```js
const weakrefs = new WeakMap();

function listen({ hostname, ...options }) {
  // ...
  if (signal) {
    weakrefs.set(signal, new WeakRef(listener));
    signal.addEventListener("abort", handleAbort);
  }
  // ...
}

function handleAbort(e) {
  weakrefs.get(e.target)?.deref()?.close();
}
```
(Use of primordials omitted for brevity.)

Reply:
@bnoordhuis that would still leak the `handleAbort` function reference.

Reply:
That's not a problem though, is it? The solutions outlined earlier leak because they create new closures every time, whereas the `handleAbort()` closure is created only once (and at the top-level so it's going to stay around forever anyway.)

Reply:
@bnoordhuis is the `handleAbort` reference still going to be in the listeners array? 

The "interesting" case here is a long-lived signal (imagine a signal owned by a controller that gets aborted on SIGINT for graceful shutdown) that gets registered often (imagine in this case you start a server on every request and do something).

Since this is `Deno.listen` it's not very likely - but that's the case we guarded against in Node